django/db/models/sql/compiler.py

```diff
@@ -137,7 +137,9 @@ def collapse_group_by(self, expressions, having):
         if self.connection.features.allows_group_by_pk:
             # The logic here is: if the main model's primary key is in the
             # query, then set new_expressions to that field. If that happens,
-            # then also add having expressions to group by.
+            # then also add having expressions to a group by and add
+            # expressions associated with a tables that don't have primary key
```

Comment:
Perhaps something like this might be easier to parse:
```
If that happens, then also add to a GROUP BY:
1. HAVING expressions
2. expressions associated with tables that don't have a primary key included
   in the grouped columns
```
but actually it might be better to try to put these comments closer to the code that implements it.